File: src/Implementations/LoadJSONManager.cpp
#include "LoadJSONManager.h"
#include "GameState.h"
#include <string>
#include <fstream>
#include <vector>
#include <stdexcept>
#include <iostream>
#include "json.hpp"

using nlohmann::json;
using std::ifstream;
using std::vector;
using std::endl;
using std::cout;
using std::exception;
using std::string;

void LoadJSONManager::LoadGameState(GameState& gameState, const string& fileName) const
{
    try
    {
        ifstream gameSave(fileName);
        if (!gameSave.is_open())
        {
            cout << "Nie mozna otworzyc pliku podczas wczytywania!" << endl;
            return;
        }
        json data = json::parse(gameSave);

        int hpHuman = data["human_HP"];
        int hpComputer = data["computer_HP"];
        int full = data["magazine_full"];
        int empty = data["magazine_empty"];
        int bulletCount = data["magazine_bullet_count"];
        int Starter = data["Starter"];
        int difficulty = data["difficulty"];
        int choice = data["choice"];
        int shooter = data["shooter"];
        int target = data["target"];
        int item = data["item"];
        int damage = data["damage"];
        bool wereHandCuffsUsed = data["WereHandCuffsUsed"];

        gameState.human.SetHP(hpHuman);
        gameState.computer.SetHP(hpComputer);
        gameState.magazine.SetFull(full);
        gameState.magazine.SetEmpty(empty);
        gameState.magazine.SetBulletCount(bulletCount);
        gameState.gameStateManager.SetStarter(Starter);
        gameState.ai.SetDifficulty(difficulty);
        gameState.gameStateManager.SetChoice(choice);
        gameState.gameStateManager.SetShooter(shooter);
        gameState.gameStateManager.SetTarget(target);
        gameState.gameStateManager.SetItem(item);
        gameState.gameStateManager.SetDamage(damage);
        gameState.gameStateManager.SetStateOfHandCuffs(wereHandCuffsUsed);

        vector<int> humanInventory;
        for (int i : data["human_inventory"]) 
        {
            humanInventory.push_back(i);
        }
        gameState.human.SetInventory(humanInventory);

        gameState.human.SetSaws(data["human_Saws"]);
        gameState.human.SetBeers(data["human_Beers"]);
        gameState.human.SetMagnifiers(data["human_Magnifiers"]);
        gameState.human.SetInverters(data["human_Inverters"]);
        gameState.human.SetHandCuffs(data["human_HandCuffs"]);
        gameState.human.SetCellPhones(data["human_CellPhones"]);

        vector<int> computerInventory;
        for (int i : data["computer_inventory"]) 
        {
            computerInventory.push_back(i);
        }
        gameState.computer.SetInventory(computerInventory);

        gameState.computer.SetSaws(data["computer_Saws"]);
        gameState.computer.SetBeers(data["computer_Beers"]);
        gameState.computer.SetMagnifiers(data["computer_Magnifiers"]);
        gameState.computer.SetInverters(data["computer_Inverters"]);
        gameState.computer.SetHandCuffs(data["computer_HandCuffs"]);
        gameState.computer.SetCellPhones(data["computer_CellPhones"]);

        vector<int> newMagazine;
        for (int i = 0; i < bulletCount; i++)
        {
            newMagazine.push_back(data["magazine"][i]);
        }
        gameState.magazine.SetMagazine(newMagazine);
    }
    catch (const json::exception& e)
    {
        cout << "Blad pliku JSON: " << e.what() << endl;
    }
    catch (const exception& e)
    {
        cout << "Blad wczytywania: " << e.what() << endl;
    }
}

File: C#/src/Implementations/LoadJSONManager.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

public class LoadJSONManager : ILoadManager
{
    public void LoadGameState(GameState gameState, string fileName)
    {
        try
        {
            string saveDir;

            if (fileName.Contains("autosave"))
            {
                saveDir = SaveConfig.GetAutoSaveDirectory();
            }
            else
            {
                saveDir = SaveConfig.GetSaveDirectory();
            }

            string fullPath = saveDir + "/" + fileName;

            using (StreamReader gameSave = new StreamReader(fullPath))
            {
                string jsonContent = gameSave.ReadToEnd();
                JObject data = JObject.Parse(jsonContent);

                int hpHuman = (int)data["human_HP"];
                int hpComputer = (int)data["computer_HP"];
                int full = (int)data["magazine_full"];
                int empty = (int)data["magazine_empty"];
                int bulletCount = (int)data["magazine_bullet_count"];
                int Starter = (int)data["Starter"];
                int difficulty = (int)data["difficulty"];
                int choice = (int)data["choice"];
                int shooter = (int)data["shooter"];
                int target = (int)data["target"];
                int item = (int)data["item"];
                int damage = (int)data["damage"];
                bool wereHandCuffsUsed = (bool)data["WereHandCuffsUsed"];

                gameState.human.SetHP(hpHuman);
                gameState.computer.SetHP(hpComputer);
                gameState.magazine.SetFull(full);
                gameState.magazine.SetEmpty(empty);
                gameState.magazine.SetBulletCount(bulletCount);
                gameState.gameStateManager.SetStarter(Starter);
                gameState.ai.SetDifficulty(difficulty);
                gameState.gameStateManager.SetChoice(choice);
                gameState.gameStateManager.SetShooter(shooter);
                gameState.gameStateManager.SetTarget(target);
                gameState.gameStateManager.SetItem(item);
                gameState.gameStateManager.SetDamage(damage);
                gameState.gameStateManager.SetStateOfHandCuffs(wereHandCuffsUsed);

                List<int> humanInventory = new List<int>();
                foreach (int i in data["human_inventory"])
                {
                    humanInventory.Add(i);
                }
                gameState.human.SetInventory(humanInventory);

                gameState.human.SetSaws((int)data["human_Saws"]);
                gameState.human.SetBeers((int)data["human_Beers"]);
                gameState.human.SetMagnifiers((int)data["human_Magnifiers"]);
                gameState.human.SetInverters((int)data["human_Inverters"]);
                gameState.human.SetHandCuffs((int)data["human_HandCuffs"]);
                gameState.human.SetCellPhones((int)data["human_CellPhones"]);

                List<int> computerInventory = new List<int>();
                foreach (int i in data["computer_inventory"])
                {
                    computerInventory.Add(i);
                }
                gameState.computer.SetInventory(computerInventory);

                gameState.computer.SetSaws((int)data["computer_Saws"]);
                gameState.computer.SetBeers((int)data["computer_Beers"]);
                gameState.computer.SetMagnifiers((int)data["computer_Magnifiers"]);
                gameState.computer.SetInverters((int)data["computer_Inverters"]);
                gameState.computer.SetHandCuffs((int)data["computer_HandCuffs"]);
                gameState.computer.SetCellPhones((int)data["computer_CellPhones"]);

                List<int> newMagazine = new List<int>();
                for (int i = 0; i < bulletCount; i++)
                {
                    newMagazine.Add((int)data["magazine"][i]);
                }
                gameState.magazine.SetMagazine(newMagazine);
            }
        }
        catch (Newtonsoft.Json.JsonException e)
        {
            Console.WriteLine("Blad pliku JSON: " + e.Message);
        }
        catch (Exception e)
        {
            Console.WriteLine("Blad wczytywania: " + e.Message);
        }
    }
}